From 2b221d65fda1d62a006a22510531bf1228f2746e Mon Sep 17 00:00:00 2001
From: Jiffin Tony Thottan <jthottan@redhat.com>
Date: Fri, 26 Jun 2015 19:05:10 +0530
Subject: [PATCH 2/3] FSAL_GLUSTER : Ignore dead links for acl fetch operation

When nfs4_getfacl call hits the deadlink it will return no entry error
So all further operation based on acl_fetch will fail.To avoid that just
ignore deadlinks.

Change-Id: Id7d3983138e82f4cb67653d49978a7d0407e3ee3
Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>
---
 src/FSAL/FSAL_GLUSTER/gluster_internal.c | 11 ++++++++---
 src/FSAL/FSAL_GLUSTER/handle.c           |  4 ++++
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/FSAL/FSAL_GLUSTER/gluster_internal.c b/src/FSAL/FSAL_GLUSTER/gluster_internal.c
index 97391a3..2ca6f93 100644
--- a/src/FSAL/FSAL_GLUSTER/gluster_internal.c
+++ b/src/FSAL/FSAL_GLUSTER/gluster_internal.c
@@ -277,9 +277,14 @@ int construct_handle(struct glusterfs_export *glexport, const struct stat *sb,
 
 	if (FSAL_IS_ERROR(status)) {
 		/* TODO: Is the error appropriate */
-		errno = EINVAL;
-		gsh_free(constructing);
-		return -1;
+
+		/* For dead links , we should not return error */
+		if (!(constructing->handle.attributes.type == SYMBOLIC_LINK
+					&& status.minor == ENOENT)) {
+			errno = EINVAL;
+			gsh_free(constructing);
+			return -1;
+		}
 	}
 
 	constructing->glhandle = glhandle;
diff --git a/src/FSAL/FSAL_GLUSTER/handle.c b/src/FSAL/FSAL_GLUSTER/handle.c
index 3de5788..7a7083e 100644
--- a/src/FSAL/FSAL_GLUSTER/handle.c
+++ b/src/FSAL/FSAL_GLUSTER/handle.c
@@ -690,6 +690,10 @@ static fsal_status_t getattrs(struct fsal_obj_handle *obj_hdl)
 	status = glusterfs_get_acl(glfs_export, objhandle->glhandle,
 				   &buffxstat, fsalattr);
 
+	/* for dead links we should not return error */
+	if (obj_hdl->type == SYMBOLIC_LINK && status.minor == ENOENT)
+		status = fsalstat(ERR_FSAL_NO_ERROR, 0);
+
  out:
 #ifdef GLTIMING
 	now(&e_time);
-- 
2.1.0

